packages/firestore/src/remote/persistent_stream.ts

```diff
@@ -253,17 +265,22 @@ export abstract class PersistentStream<
    * be in a !isStarted() state, requiring the caller to start the stream again
    * before further use.
    *
-   * Only streams that are in state 'Open' can be marked idle, as all other
-   * states imply pending network operations.
+   * Only streams that are in state 'Open' or re-trying can be marked idle, as
+   * all other states imply pending network operations.
+   *
+   * @param listener the listener to call on close event.
    */
-  markIdle(): void {
+  markIdle(listener?: ListenerType): void {
+    if (isNullOrUndefined(listener)) {
+      listener = this.listener;
+    }
     // Starts the idle time if we are in state 'Open' and are not yet already
     // running a timer (in which case the previous idle timeout still applies).
-    if (this.isOpen() && this.inactivityTimerPromise === null) {
+    if (this.canTimeout() && this.inactivityTimerPromise === null) {
       this.inactivityTimerPromise = this.queue.enqueueAfterDelay(
         this.idleTimerId,
         IDLE_TIMEOUT_MS,
-        () => this.handleIdleCloseTimer()
+        () => this.handleIdleCloseTimer(this.listener)
```

Comment:
In order to recover this.listener in the callback, we need to pass it into the callback. It used to assume Idle timer is only for a running (online) stream. Now we allow Idle timer for a re-trying stream (which is a retry after an offline failure and presumable also offline). An offlined stream does not have listener property as we only keep it for a online stream. In other words, in callback, this.listener is valid for an online stream while this.listener is null for an offline stream. Passing in this.listener is the minimal change we can make without breaking more assumption in code. By minimal I mean change the idle timer logic here instead of let an offline stream keep this.listener where idle timer is only triggered for a small fraction of usage of watch streams i.e. majority watch streams are not idle.

Reply:
You're passing in `listener` but then using `this.listener` here. I assume that must be a mistake?